src/smolagents/agents.py

```diff
@@ -1412,6 +1456,156 @@ def _step_stream(self, memory_step: ActionStep) -> Generator[Any]:
         memory_step.action_output = output
         yield output if is_final_answer else None
 
+    def _json_step(self, memory_step: ActionStep) -> Generator[Any]:
```

Comment:
The implementation of `_json_step` duplicates most of the code in `_step`. This should be optimized to reduce redundancy.